main.cpp

```diff
@@ -1,239 +1,239 @@
+     sigemptyset(&tcb->env->__saved_mask);
+ 
+     // We now have one more thread!
+     num_threads++;
+ }
+ 
+ void uthread_yield(){
+ 
+     // Save execution state.
+     int ret_val = sigsetjmp(threads[current_thread_id].env,1);
+     if (ret_val == 1) return;
+ 
+     // Choose the next thread to execute.
+     int id = current_thread_id;
+     do {
+         id = (id + 1) % num_threads;
+         if(id == current_thread_id){
+             printf("infinite loop!\n");
+         }
+     } while (threads[id].complete);
+ 
+     // Execute that thread.
+     current_thread_id = id;
+     siglongjmp(threads[current_thread_id].env,1);
+ }
+ 
+ int uthread_self(){
+     return current_thread_id;
+ }
+ 
+ void start(){
+     current_thread_id = 0;
+     siglongjmp(threads[current_thread_id].env,1);
+ }
+ 
+ ////////////////////////////////////
+ //////////////////////////////////// Test Cases
+ ////////////////////////////////////
+ 
+ void* uthread_test_function(void* arg){
+     assert((long)arg == 10);
+     for(int i=0;i<100;i++){
+         printf("%d\n", i);
+     }
+     return 0;
+ }
+ void* uthread_yield_test_function(void* arg){
+     for(int i=0;true;i++){
+         if(i % 500000 == 0){
+             uthread_yield();
+         }
+         printf("%d\n", i);
+     }
+     return 0;
+ }
+ 
+ void* uthread_self_test_function(void* arg){
+     static int id = 0;
+     assert(uthread_self()==id);
+     id++;
+     return nullptr;
+ }
+ 
+ void* f(void * arg){
+     assert((long)arg == 10);
+     printf("argument: %ld\n", (long)arg);
+     int i=0;
+     while(1) {
+         ++i;
+         printf("in f (%d)\n",i);
+         if(i % 5 == 0){
+         }
+         if (i % 3 == 0) {
+             printf("f: switching\n");
+             uthread_yield();
+         }
+         usleep(SECOND);
+     }
+     return 0;
+ }
+ 
+ void* g(void * arg){
+     printf("&arg: %p\n", &arg);
+     printf("arg: 0x%lx\n", (long)arg);
+     int i=0;
+     while(1){
+         ++i;
+         printf("in g (%d)\n",i);
+         if (i % 5 == 0) {
+             printf("g: switching\n");
+             uthread_yield();
+         }
+         usleep(SECOND);
+     }
+     return 0;
+ }
+ 
+ void test_uthread_create(){
+     uthread_create(uthread_test_function, (void*)10l);
+ }
+ 
+ // Create 10 threads, and check that they alternate between eachother
+ void test_thread_yield(){
+     uthread_create(uthread_yield_test_function, nullptr);
+     uthread_create(uthread_yield_test_function, nullptr);
+     uthread_create(uthread_yield_test_function, nullptr);
+     uthread_create(uthread_yield_test_function, nullptr);
+     uthread_create(uthread_yield_test_function, nullptr);
+ }
+ 
+ void test_thread_self(){
+     uthread_create(uthread_self_test_function, nullptr);
+ }
+ 
+ void* yield_wrapper(void* arg){
+     test_thread_yield();
+ }
+ 
+ int main(){
+     test_thread_self();
+     test_uthread_create();
+     uthread_create(yield_wrapper, nullptr);
```

Comment:
This code is a little freaky because we are running a 2 tests, then scheduling running the thread_yield tests with thread_create, effectively testing that we can schedule threads within a thread.

Reply:
So this “code review” is kind of hacky: I opened a pull request shifting each line of code by 1 character, then I commented on the differences by clicking “view change” from the PR, then clicking on individual lines allows you to comment on them. You can do the same by reviewing the pull request.

Reply:
@kstowe